### PR TITLE
gh-107073: Make PyObject_VisitManagedDict() public

### PR DESCRIPTION
**Pull request by vstinner**
Make PyObject_VisitManagedDict() and PyObject_ClearManagedDict() functions public in Python 3.13 C API.

* Rename _PyObject_VisitManagedDict() to PyObject_VisitManagedDict().
* Rename _PyObject_ClearManagedDict() to PyObject_ClearManagedDict().
* Document these functions.

**EDIT**: Target Python 3.13, not Python 3.12.

* Issue: gh-107073

----
:books: Documentation preview :books:: https://cpython-previews--108763.org.readthedocs.build/

### CODE REVIEW

**Comment by gvanrossum** [comment]
According to NEWS when these were added, they were unstable. Presumably because they mess with internals that might change. Then shouldn't they be named PyUnstable_...? Or, @markshannon have you changed your mind about their status?

**Comment by gvanrossum** [comment]
It's too late for 3.12 right?

**Comment by vstinner** [comment]
> It's too late for 3.12 right?

That's a question for @Yhg1s, Python 3.12 release manager. This change adds 2 functions to the public C API.

I would prefer to fix the API in Python 3.12, but if it's too late, it's too late :-)

**Comment by vstinner** [comment]
If it's too late for Python 3.12, that's ok. The private functions can be used. But I will update my PR to document the change in What's new in Python 3.13 instead ;-) Just tell what you think about this one @Yhg1s ;-)

**Comment by Yhg1s** [comment]
No, I think it's too late for 3.12.

**Comment by vstinner** [comment]
@gvanrossum:
> According to NEWS when these were added, they were unstable. Presumably because they mess with internals that might change. Then shouldn't they be named PyUnstable_...? Or, @markshannon have you changed your mind about their status?

What do you mean when you say that ``_PyObject_VisitManagedDict()`` and ``_PyObject_ClearManagedDict()`` are *unstable*? Are you referring to the risk of bugs? Or the risk that the API may change soon?

Are you referring to the What's New in Python 3.12 doc?

```
* Extension classes wanting to add a ``__dict__`` or weak reference slot
  should use :c:macro:`Py_TPFLAGS_MANAGED_DICT` and
  :c:macro:`Py_TPFLAGS_MANAGED_WEAKREF` instead of ``tp_dictoffset`` and
  ``tp_weaklistoffset``, respectively.
  The use of ``tp_dictoffset`` and ``tp_weaklistoffset`` is still
  supported, but does not fully support multiple inheritance
  (:gh:`95589`), and performance may be worse.
  Classes declaring :c:macro:`Py_TPFLAGS_MANAGED_DICT` should call
  :c:func:`!_PyObject_VisitManagedDict` and :c:func:`!_PyObject_ClearManagedDict`
  to traverse and clear their instance's dictionaries.
  To clear weakrefs, call :c:func:`PyObject_ClearWeakRefs`, as before.
```

The "does not fully support multiple inheritance" part is scary. The ``PyTypeObject.tp_dictoffset`` documentation strongly suggests moving towards the ``Py_TPFLAGS_MANAGED_DICT`` API:

```
.. c:member:: Py_ssize_t PyTypeObject.tp_dictoffset

   While this field is still supported, :c:macro:`Py_TPFLAGS_MANAGED_DICT` should be
   used instead, if at all possible.

   (...)

   **Inheritance:**

   This field is inherited by subtypes. A subtype should not override this offset;
   doing so could be unsafe, if C code tries to access the dictionary at the
   previous offset.
   To properly support inheritance, use :c:macro:`Py_TPFLAGS_MANAGED_DICT`.
```

**Comment by vstinner** [comment]
@Yhg1s:
> No, I think it's too late for 3.12.

Ok, the release manager spoke, I re-targeted my PR to Python 3.13 :-) It's ok to use the private API in Python 3.12.

**Comment by gvanrossum** [comment]
> What do you mean when you say that `_PyObject_VisitManagedDict()` and `_PyObject_ClearManagedDict()` are _unstable_? Are you referring to the risk of bugs? Or the risk that the API may change soon?

It's literally what the [NEWS entry for 3.12a1](https://github.com/python/cpython/blob/main/Misc/NEWS.d/3.12.0a1.rst) says:
```
.. date: 2022-07-25-15-54-27
.. gh-issue: 92678
.. nonce: ziZpxz
.. section: C API

Adds unstable C-API functions ``_PyObject_VisitManagedDict`` and
``_PyObject_ClearManagedDict`` to allow C extensions to allow the VM to
manage their object's dictionaries.
```

**Comment by vstinner** [comment]
@gvanrossum
> Or, @markshannon have you changed your mind about their status?

So @gvanrossum is worried by the Changelog entry adding these 2 APIs as "unstable".

@markshannon: You added these functions 1 year ago in Python 3.12 as commit de388c0a7b71c094d36ce40fecef87bdbb8a87d3. Do you still consider these APIs as "unstable"? Are you ok to make them public? Well, see the issue and previous comments for the rationale ;-)

**Comment by vstinner** [comment]
For me, it's unfortunate that Python 3.12 broke type inheritance with the new "managed dict" change, but doesn't offer a public API to switch to it, while the documentation strongly advices using this new API. That's why I asked to make this API public in Python 3.12.

But this issue is waiting for 1 month, and Python 3.12 release is following its process towards the final release in one month.

It's unclear to me if inheritance was broken in Python 3.11 or 3.12. I recall that pybind11 got a regression related to that, and as I wrote. It now uses a the private ``_PyObject_GetDictPtr()`` API to visit and clear the managed dict: https://github.com/python/cpython/issues/107073#issuecomment-1646675371 So it worked around the lack _PyObject_VisitManagedDict() and _PyObject_ClearManagedDict() in Python 3.11.

Well, even if I dislike the implementation, it's good to know that *there is a way* to implement PyObject_VisitManagedDict() and PyObject_ClearManagedDict() on Python 3.11.

It also means that there is a need for public API for that, since pybind11 already uses it :-)

**Comment by vstinner** [comment]
> I recall that pybind11 got a regression related to that, and as I wrote. It now uses a the private _PyObject_GetDictPtr() API to visit and clear the managed dict: https://github.com/python/cpython/issues/107073#issuecomment-1646675371

If we decide adding ``PyObject_VisitManagedDict()`` to the public C API, what I can is to provide a public PyObject_VisitManagedDict() function for Python 3.11 and 3.12 in the https://github.com/python/pythoncapi-compat/ project using ``_PyObject_GetDictPtr()``. So projects can use the new public API since Python 3.11 :-)

So I don't have to be pushy with our busy Python 3.12 release manager to get it merged after beta1 (after rc1!!!) ;-)

**Comment by vstinner** [comment]
@dean0x7d @wjakob: Hi, you modified pybin11 to travere and clear the dictionary in pybind11 ([commit](https://github.com/pybind/pybind11/commit/08cbe8dfed1f245a17daf91babeb950a2aac82c9)) with such code:

```
/// dynamic_attr: Allow the garbage collector to traverse the internal instance `__dict__`.
extern "C" inline int pybind11_traverse(PyObject *self, visitproc visit, void *arg) {
    PyObject *&dict = *_PyObject_GetDictPtr(self);
    Py_VISIT(dict);
    return 0;
}

/// dynamic_attr: Allow the GC to clear the dictionary.
extern "C" inline int pybind11_clear(PyObject *self) {
    PyObject *&dict = *_PyObject_GetDictPtr(self);
    Py_CLEAR(dict);
    return 0;
}
```

I see that pybind11 uses Py_TPFLAGS_MANAGED_DICT on Python 3.11. Would it help you to have public function PyObject_VisitManagedDict() and PyObject_ClearManagedDict()? Do you have an opinion on the API? Is it good?

**Comment by wjakob** [comment]
The proposal looks fine to me.

A tangent: a more general solution would be to expose the dictionary in a public (perhaps even stable) API interface, but not like the existing mechanism. `_PyObject_GetDictPtr` is problematic because it exposes CPython implementation details. If the signature was `PyObject_GetDict()` instead (returning `PyObject*` and not `PyObject **`) then that is enough information to find and break reference cycles. Those functions could then be implemented as

```
/// dynamic_attr: Allow the garbage collector to traverse the internal instance `__dict__`.
extern "C" inline int pybind11_traverse(PyObject *self, visitproc visit, void *arg) {
    PyObject *dict = PyObject_GetDict(self);
    Py_VISIT(dict);
    return 0;
}

/// dynamic_attr: break a reference cycle involving this object
extern "C" inline int pybind11_clear(PyObject *self) {
    PyObject *dict = PyObject_GetDict(self);
    PyDict_Clear(dict);
    return 0;
}
```

**Comment by vstinner** [comment]
In short, you ask to make ``_PyObject_GetDict()`` public? Yeah, I am curious if it would make sense to make this function public as well. By the way, this  function is documented as: "**should be treated as part of the public API**"!

```c
/* Helper to get a pointer to an object's __dict__ slot, if any.
 * Creates the dict from inline attributes if necessary.
 * Does not set an exception.
 *
 * Note that the tp_dictoffset docs used to recommend this function,
 * so it should be treated as part of the public API.
 */
```

**Comment by wjakob** [comment]
@vstinner Ah, interesting—I did not know about this function. The fact that it creates the dictionary if non-existent makes it inappropriate for what I had suggested (for use in garbage collection hooks).

**Comment by vstinner** [comment]
> @vstinner Ah, interesting—I did not know about this function. The fact that it creates the dictionary if non-existent makes it inappropriate for what I had suggested (for use in garbage collection hooks).

Can we discuss exposing _PyObject_GetDict() separetly? Exposing just PyObject_VisitManagedDict() and PyObject_ClearManagedDict() already looks controversial :-)

**Comment by vstinner** [comment]
@markshannon @gvanrossum @encukou: this issue is languishing since July 22. I plan to merge my PR next Friday. If you need more time for review, please speak up.

Once this PR will be merged, I consider adding a compatible layer for Python 3.11 and 3.12 in pythoncapi-compat. Later, I consider making _PyObject_GetDict() public, but apparently this function needs to be adjusted before being made public.

**Comment by wjakob** [comment]
I understand that exposing these functions would be helpful to fix garbage collection of managed dictionaries in extension libraries like pybind11. But unless there is a big rush to push this out for a release (which I don't think is the case here, right?), isn't it better to refine and have a perfect API? In this case, it somehow seems suboptimal to expose two highly specific public functions as public API if they are likely to be subsumed by a more general one later one.

**Comment by vstinner** [comment]
> it somehow seems suboptimal to expose two highly specific public functions as public API if they are likely to be subsumed by a more general one later one.

I don't think that anyone has a plan to replace these two convenient functions with something like _PyObject_GetDict().

_PyObject_GetDict() cannot be used directly in traverse and clear functions, IMO it's less convenient to use. See my Doc/c-api/typeobj.rst changes in this PR: PyObject_VisitManagedDict() and PyObject_ClearManagedDict() are convenient to use.

For me, _PyObject_GetDict() fits other use cases, no?

**Comment by gvanrossum** [comment]
Both Mark and I are on vacation this week. Please wait.

**Comment by vstinner** [comment]
> Both Mark and I are on vacation this week. Please wait.

So, any update on these functions? Should they be exposed? Is it the new official way to have a dict in an object?

Or should we burry them deeper in the internal C API?

**Comment by gvanrossum** [comment]
Pinging @markshannon

**Comment by markshannon** [comment]
I've commented on the issue on my thinking as to why we should add these to the public API.

File: Doc/c-api/typeobj.rst
Comment (markshannon) [inline Doc/c-api/typeobj.rst:1374]: This doesn't seem relevant `PyObject_VisitManagedDict`, but no harm in adding it.

**Review by markshannon — APPROVED** [review]

File: Doc/c-api/typeobj.rst
Comment (vstinner) [reply Doc/c-api/typeobj.rst:1374]: This doc was forgotten since Python 3.8, it requires to fix an important bug.

**Comment by vstinner** [comment]
@vstinner:
> Now we should visit the _PyObject_GetDict() case, hum.

Oh, @markshannon replied that we should not expose it: https://github.com/python/cpython/issues/107073#issuecomment-1742624694

@wjakob: If you want, you can now use the pythoncapi-project to use PyObject_VisitManagedDict() and PyObject_ClearManagedDict() on Python 3.11 and Python 3.12, implemented using _PyObject_GetDictPtr(), see: https://github.com/python/pythoncapi-compat/commit/a5943544e2301aa6f5c026b48d84d32e01d9ebda

**Comment by wjakob** [comment]
@vstinner  — those two functions would be usable for pybind11 (@rwgk, @henryiii).

For nanobind, I'm still using old-style dictionaries built into each instance :-/. I don't this functionality is sufficiently exposed through limited API / stable ABI interfaces to be able to switch to managed dicts. Is it the plan that stable ABI interfaces can also use them?

**Pull request merged**
